python/GafferUITest/GraphGadgetTest.py

```diff
@@ -1406,5 +1408,336 @@ def testChangeNodeGadgetForUnviewedNode( self ) :
+		box["BoxOut1"]["passThrough"].setInput( None )
+		plugs, nodes = GafferUI.GraphGadget._activePlugsAndNodes( s["add2"]["sum"], c, None )
+		self.assertEqual( set( plugs ), set( [ s["add2"]["op1"], box["sum1"], box["BoxOut1"]["__out"], box["BoxOut1"]["__switch"]["in"]["in0"] ] ) )
+		self.assertEqual( set( nodes ), set( [ box, s["add2"], box["BoxOut1"]["__switch"], box["BoxOut1"] ] ) )
+
+
+		# Test Loop
+
+		s["loopSwitch"] = Gaffer.Switch()
+		s["loopSwitch"].setup( Gaffer.FloatPlug() )
+		for i in range( 5 ):
+			s["loopSwitchIn%i" % i ] = GafferTest.AddNode()
+			s["loopSwitchIn%i" % i ]["op1"].setValue( (i+1) * 10 ** i )
+			s["loopSwitch"]["in"][-1].setInput( s["loopSwitchIn%i" % i ]["sum"] )
+
+		s["loopSwitch"]["expr"] = Gaffer.Expression()
+		s["loopSwitch"]["expr"].setExpression( 'parent["index"] = context.get("loop:index", 4)', "python" )
+
+		s["start"] = GafferTest.AddNode()
+		s["start"]["op1"].setValue( 900000 )
+
+		s["loop"] = Gaffer.Loop()
+		s["loop"].setup( Gaffer.FloatPlug() )
+		s["loop"]["in"].setInput( s["start"]["sum"] )
+
+		s["merge"] = GafferTest.AddNode()
+		s["merge"]["op1"].setInput( s["loop"]["previous"] )
+		s["loop"]["next"].setInput( s["merge"]["sum"] )
+		s["merge"]["op2"].setInput( s["loopSwitch"]["out"] )
+
+		s["loop"]["iterations"].setValue( 4 )
+
+		self.assertEqual( s["loop"]["out"].getValue(), 904321.0 )
+
+		# We don't track through the loop for every separate loop context, we just fall back to a naive
+		# traversal that takes every input
+		plugs, nodes = GafferUI.GraphGadget._activePlugsAndNodes( s["loop"]["out"], c, None )
+		self.assertEqual( set( nodes ), set( [ s["loop"], s["start"], s["merge"], s["loopSwitch"], s["loopSwitch"]["expr"] ] + [ s["loopSwitchIn%i"%i] for i in range(5) ] ) )
+
+		# Same deal if we start from a node in the middle of the loop iteration
+		plugs, nodes = GafferUI.GraphGadget._activePlugsAndNodes( s["merge"]["sum"], c, None )
+		self.assertEqual( set( nodes ), set( [ s["loop"], s["start"], s["merge"], s["loopSwitch"], s["loopSwitch"]["expr"] ] + [ s["loopSwitchIn%i"%i] for i in range(5) ] ) )
+
+	def testActivePlugsAndNodesCancellation( self ) :
+
+		def testThreadFunc( plug, canceller ):
+			try:
+				# TODO - should probably omit canceller from this API, and just take a context?
+				GafferUI.GraphGadget._activePlugsAndNodes( plug, Gaffer.Context( Gaffer.Context(), canceller ), canceller )
+			except:
+				pass # We're expecting to be cancelled
+
+		s = Gaffer.ScriptNode()
+
+		# Test a computation that supports cancellation
+
+		s["add"] = GafferTest.AddNode()
+
+		s["addExpr"] = Gaffer.Expression()
+		s["addExpr"].setExpression( inspect.cleandoc(
+			"""
+			import time
+			import Gaffer
+			for i in range( 50 ):
+				time.sleep( 0.2 )
+				IECore.Canceller.check( Gaffer.Context.current().canceller() )
+
+			parent["add"]["enabled"] = True
+			"""
+		) )
+
+		canceller = IECore.Canceller()
+		thread = threading.Thread(
```

Comment:
Doesn't really matter in the tests, but in real life it's definitely preferable to use `ParallelAlgo.runInBackgroundThread()`, so we use one of the pre-existing TBB threads. Otherwise we make a bunch of thread-local caches for the new thread, and I don't think `enumerable_thread_specific` cleans them up when the thread ends?

Reply:
Addressed in 3e0ab6d.